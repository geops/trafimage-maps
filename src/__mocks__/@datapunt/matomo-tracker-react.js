const {
  MatomoProvider: MP,
  MatomoContext: MC,
  useMatomo: uM,
  createInstance: cI,
} = require('@datapunt/matomo-tracker-react');

export const createInstance = jest.fn((params) => {
  const inst = cI(params);
  inst.pushInstruction = jest.fn();
  return inst;
});
export const useMatomo = uM;
export const MatomoContext = MC;
export const MatomoProvider = MP;
