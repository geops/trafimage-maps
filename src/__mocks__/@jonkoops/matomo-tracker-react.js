const {
  MatomoProvider: MP,
  MatomoContext: MC,
  useMatomo: uM,
  createInstance: cI,
} = require("@jonkoops/matomo-tracker-react");

const createInstanceTmp = jest.fn((params) => {
  const inst = cI(params);
  inst.pushInstruction = jest.fn();

  // Store the lastinstance for testing
  createInstanceTmp.lastInstance = inst;
  return inst;
});

export const createInstance = createInstanceTmp;
export const useMatomo = uM;
export const MatomoContext = MC;
export const MatomoProvider = MP;
