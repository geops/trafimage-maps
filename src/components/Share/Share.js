import React from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { TiImage, TiSocialFacebook, TiSocialTwitter } from 'react-icons/ti';
import { FaEnvelope } from 'react-icons/fa';
import CanvasSaveButton from 'react-spatial/components/CanvasSaveButton';
import BlankLink from 'react-spatial/components/BlankLink';
import SharePermalinkButton from '../SharePermalinkButton';

const socialShareConfig = [
  {
    url: 'mailto:?body={url}',
    title: 'Per Email versenden',
    icon: <FaEnvelope focusable={false} />,
    className: 'ta-mail-icon',
  },
  {
    url: '//www.facebook.com/sharer.php?u={url}',
    title: 'Auf Facebook teilen',
    icon: <TiSocialFacebook focusable={false} />,
    className: 'ta-facebook-icon',
  },
  {
    url: '//twitter.com/intent/tweet?url={url}',
    title: 'Auf Twitter teilen',
    icon: <TiSocialTwitter focusable={false} />,
    className: 'ta-twitter-icon',
  },
];

const renderConf = (conf, t) => (
  <div className={conf.className} key={conf.title}>
    <BlankLink href={conf.url} title={t(conf.title)}>
      {conf.icon}
    </BlankLink>
  </div>
);

const Share = () => {
  const map = useSelector(state => state.app.map);
  const { t } = useTranslation();
  const config = [...socialShareConfig];

  for (let i = 0; i < config.length; i += 1) {
    if (config[i].url) {
      config[i].url = config[i].url.replace('{url}', window.location.href);
    }
  }

  return (
    <div className="wkp-share">
      <SharePermalinkButton />
      {renderConf(config[0], t)}
      <CanvasSaveButton title={t('Karte als Bild speichern')} map={map}>
        <TiImage focusable={false} />
      </CanvasSaveButton>
      {renderConf(config[1], t)}
      {renderConf(config[2], t)}
    </div>
  );
};

export default React.memo(Share);
