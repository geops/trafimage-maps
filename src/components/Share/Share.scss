.wkp-share {
  width: 100%;
  display: flex;

  > div {
    width: 30px;
    height: 30px;
    margin: 10px 0 10px 20px;
    display: flex;
    align-items: center;
    justify-content: space-around;
  }

  .ta-mail-icon {
    a {
      display: flex;
      align-items: center;
      justify-content: space-around;

      svg {
        font-size: 23px;
        margin-bottom: 2px;
      }
    }
  }

  .wkp-share-permalink-button > div[role='button'] {
    width: 30px;
    height: 30px;
    display: flex;
    align-items: center;
    justify-content: space-around;

    svg {
      font-size: 20px;
    }
  }

  .ta-facebook-icon {
    margin: 10px 0 10px 15px;
    font-size: 30px;
  }

  .ta-twitter-icon,
  .tm-canvas-save-button {
    font-size: 30px;
  }

  .tm-canvas-save-button {
    font-size: 28px;
  }
}
