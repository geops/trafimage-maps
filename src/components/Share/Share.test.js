import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { Map, View } from "ol";
import { ThemeProvider } from "@mui/material";
import { MatomoProvider } from "@jonkoops/matomo-tracker-react";
import theme from "../../themes/default";
import Share from ".";

describe("Share", () => {
  let store;
  delete global.window.location;
  global.window = Object.create(window);
  global.window.location = { hostname: "wkp-dev.foo" };
  beforeEach(() => {
    store = global.global.mockStore({
      map: {},
      app: {
        map: new Map({ view: new View({}) }),
        activeTopic: {
          key: "test",
        },
        language: "de",
        appBaseUrl: "https://maps.trafimage.ch",
      },
    });
  });

  test("should match snapshot.", () => {
    const { container } = render(
      <MatomoProvider value={{}}>
        <ThemeProvider theme={theme}>
          <Provider store={store}>
            <Share />
          </Provider>
        </ThemeProvider>
      </MatomoProvider>,
    );
    expect(container.querySelectorAll("button").length).toBe(2);
    expect(container.innerHTML).toMatchSnapshot();
  });

  describe("should send track event on click", () => {
    let matomo;
    let container;

    beforeEach(() => {
      matomo = {
        pushInstruction: jest.fn(),
        trackPageView: jest.fn(),
        trackEvent: jest.fn(),
      };
      const wrapper = render(
        <MatomoProvider value={matomo}>
          <ThemeProvider theme={theme}>
            <Provider store={store}>
              <Share />
            </Provider>
          </ThemeProvider>
        </MatomoProvider>,
      );
      container = wrapper.container;
      window.digitalDataLayer = [];
    });

    test("on permalink button", () => {
      fireEvent.click(container.querySelector(".wkp-permalink-bt button"));
      expect(matomo.trackEvent).toBeCalledWith({
        action: "clickSharePermalink",
        category: "test",
      });
      expect(window.digitalDataLayer[0].event.eventInfo.variant).toMatch(
        /Permalink erstellen/i,
      );
    });

    test("on mail button", () => {
      fireEvent.click(container.querySelector(".ta-mail-icon a"));
      expect(matomo.trackEvent).toBeCalledWith({
        action: "clickShareMail",
        category: "test",
      });
      expect(window.digitalDataLayer[0].event.eventInfo.variant).toMatch(
        /Per Email versenden/i,
      );
    });

    test("on download button", () => {
      fireEvent.click(container.querySelector(".rs-canvas-save-button"));
      expect(matomo.trackEvent).toBeCalledWith({
        action: "clickShareDownload",
        category: "test",
      });
      expect(window.digitalDataLayer[0].event.eventInfo.variant).toMatch(
        /PNG export/i,
      );
    });

    test("on facebook button", () => {
      fireEvent.click(container.querySelector(".ta-facebook-icon a"));
      expect(matomo.trackEvent).toBeCalledWith({
        action: "clickShareFacebook",
        category: "test",
      });
      expect(window.digitalDataLayer[0].event.eventInfo.variant).toMatch(
        /Auf Facebook teilen/i,
      );
    });

    test("on twitter button", () => {
      fireEvent.click(container.querySelector(".ta-twitter-icon a"));
      expect(matomo.trackEvent).toBeCalledWith({
        action: "clickShareTwitter",
        category: "test",
      });
      expect(window.digitalDataLayer[0].event.eventInfo.variant).toMatch(
        /Auf Twitter teilen/i,
      );
    });
  });
});
