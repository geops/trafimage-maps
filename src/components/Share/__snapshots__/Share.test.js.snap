// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share should match snapshot. 1`] = `
<div
  className="wkp-share"
>
  <div
    className="wkp-share-permalink-button"
  >
    <div
      aria-label="Permalink generieren"
      className="tm-button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
      title="Permalink generieren"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        />
      </svg>
    </div>
  </div>
  <div
    className="ta-mail-icon"
  >
    <a
      href="mailto:?body=http://localhost/"
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
      title="Per Email versenden"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
        />
      </svg>
    </a>
  </div>
  <div
    aria-label="Karte als Bild speichern"
    className="rs-canvas-save-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    scale={1}
    tabIndex={0}
  >
    <svg
      baseProfile="tiny"
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      version="1.2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Karte als Bild speichern
      </title>
      <circle
        cx="8.5"
        cy="8.501"
        r="2.5"
      />
      <path
        d="M16 10c-2 0-3 3-4.5 3s-1.499-1-3.5-1c-2 0-3.001 4-3.001 4h14.001s-1-6-3-6zM20 3h-16c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h16c1.103 0 2-.897 2-2v-12c0-1.103-.897-2-2-2zm0 14h-16v-12h16v12z"
      />
    </svg>
  </div>
  <div
    className="ta-facebook-icon"
  >
    <a
      href="//www.facebook.com/sharer.php?u=http://localhost/"
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
      title="Auf Facebook teilen"
    >
      <svg
        baseProfile="tiny"
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 10h3v3h-3v7h-3v-7h-3v-3h3v-1.255c0-1.189.374-2.691 1.118-3.512.744-.823 1.673-1.233 2.786-1.233h2.096v3h-2.1c-.498 0-.9.402-.9.899v2.101z"
        />
      </svg>
    </a>
  </div>
  <div
    className="ta-twitter-icon"
  >
    <a
      href="//twitter.com/intent/tweet?url=http://localhost/"
      rel="noopener noreferrer"
      tabIndex="0"
      target="_blank"
      title="Auf Twitter teilen"
    >
      <svg
        baseProfile="tiny"
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.89 7.012c.808-.496 1.343-1.173 1.605-2.034-.786.417-1.569.703-2.351.861-.703-.756-1.593-1.14-2.66-1.14-1.043 0-1.924.366-2.643 1.078-.715.717-1.076 1.588-1.076 2.605 0 .309.039.585.117.819-3.076-.105-5.622-1.381-7.628-3.837-.34.601-.51 1.213-.51 1.846 0 1.301.549 2.332 1.645 3.089-.625-.053-1.176-.211-1.645-.47 0 .929.273 1.705.82 2.388.549.676 1.254 1.107 2.115 1.291-.312.08-.641.118-.979.118-.312 0-.533-.026-.664-.083.23.757.664 1.371 1.291 1.841.625.472 1.344.721 2.152.743-1.332 1.045-2.855 1.562-4.578 1.562-.422 0-.721-.006-.902-.038 1.697 1.102 3.586 1.649 5.676 1.649 2.139 0 4.029-.542 5.674-1.626 1.645-1.078 2.859-2.408 3.639-3.974.784-1.564 1.172-3.192 1.172-4.892v-.468c.758-.57 1.371-1.212 1.84-1.921-.68.293-1.383.492-2.11.593z"
        />
      </svg>
    </a>
  </div>
  <div
    className="ta-draw-icon"
  >
    <div
      aria-label="Zeichnen."
      className="tm-button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
      title="Zeichnen."
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
        />
      </svg>
    </div>
  </div>
</div>
`;
